rpcs3/rpcs3qt/emu_settings.cpp

```diff
@@ -307,6 +307,86 @@ void emu_settings::EnhanceCheckBox(QCheckBox* checkbox, emu_settings_type type)
 	});
 }
 
+void emu_settings::EnhanceDateTimeEdit(QDateTimeEdit* date_time_edit, emu_settings_type type, const QString& format, bool use_calendar, bool as_offset_from_now)
+{
+	if (!date_time_edit)
+	{
+		cfg_log.fatal("EnhanceDateTimeEdit '%s' was used with an invalid object", cfg_adapter::get_setting_name(type));
+		return;
+	}
+
+	date_time_edit->setDisplayFormat(format);
+	date_time_edit->setCalendarPopup(use_calendar);
+
+	if (as_offset_from_now)
+	{
+		bool ok_def = false;
+		const s64 def = qstr(GetSettingDefault(type)).toLongLong(&ok_def);
+		if (!ok_def)
+		{
+			cfg_log.fatal("EnhanceDateTimeEdit '%s' was used with an invalid emu_settings_type", cfg_adapter::get_setting_name(type));
+			return;
+		}
+
+		bool ok_sel = false;
+		s64 val = qstr(GetSetting(type)).toLongLong(&ok_sel);
+		if (!ok_sel)
+		{
+			cfg_log.error("EnhanceDateTimeEdit '%s' tried to set an invalid value: %d. Setting to default: %d", cfg_adapter::get_setting_name(type), val, def);
+			val = def;
+			m_broken_types.insert(type);
+			SetSetting(type, std::to_string(def));
+		}
+
+		// get the date_time value with the offset applied, and set it to the control
+		const QDateTime date_time = QDateTime::currentDateTime().addSecs(val);
+		date_time_edit->setDateTime(date_time);
+	}
+	else
+	{
+		auto str = qstr(GetSettingDefault(type));
+		const auto def = QDateTime::fromString(str, Qt::ISODate);
+		const bool ok_def = def.isValid();
+		if (!ok_def)
+		{
+			cfg_log.fatal("EnhanceDateTimeEdit '%s' was used with an invalid emu_settings_type", cfg_adapter::get_setting_name(type));
+			return;
+		}
+
+		str = qstr(GetSetting(type));
+		auto val = QDateTime::fromString(str, Qt::ISODate);
+		const bool ok_sel = val.isValid();
+		if (!ok_sel)
+		{
+			cfg_log.error("EnhanceDateTimeEdit '%s' tried to set an invalid value: %s. Setting to default: %s", \
+				cfg_adapter::get_setting_name(type), val.toString(Qt::ISODate).toStdString(), def.toString(Qt::ISODate).toStdString());
+			val = def;
+			m_broken_types.insert(type);
+			SetSetting(type, sstr(def.toString(Qt::ISODate)));
+		}
+
+		// set the date_time value to the control
+		date_time_edit->setDateTime(val);
+	}
+
+
+	connect(date_time_edit, &QDateTimeEdit::dateTimeChanged, [date_time_edit, type, as_offset_from_now, this](const QDateTime& datetime)
+	{
+		if (as_offset_from_now)
+		{
+			// offset will be applied in seconds
+			const s64 offset = QDateTime::currentDateTime().secsTo(datetime);
+			SetSetting(type, std::to_string(offset));
+			date_time_edit->clearFocus();
```

Comment:
Why is a clear focus only here?

Reply:
The QDateTimeEdit 'offset-driven update' is suppressed when the QDateTimeEdit (or its child Calendar widget) have focus.  Otherwise every second it was changing the QDateTimeEdit value which was taking the widget out of edit mode (stomping over in progress changes the user was making).
So once the value is changed, then I deliberately 'throw away' the focus, so that it returns to updating.  As item (f) in the list mentions, I'm very much open to other ways of handling this, but there didn't appear to be any 'edit in progress' indication available from the Qt framework.

Reply:
I mean: why is it not always set at the end of this lambda? why only in the offset branch?

Reply:
It's part of the 'hack' to allow the date/time to dynamically update, whilst also allowing users to change the date/time without their change being overwritten.  It's definitely not an amazing implementation of this particular portion, but I couldn't find anything to handle this in Qt.
Where the date/time is static, then it can keep the focus.  We don't update it dynamically, so we don't need to worry if the user is currently editing it or similar.
Qt doesn't appear to have any isBeingEdited() status... so hasFocus() wasn't the closest available.
It seems to work well enough for the likely use cases - occasional editing where the exact time isn't important, but going back in time by xx days / months / years is needed.